test/karma-testharness-adapter.js

```diff
@@ -0,0 +1,181 @@
+
+  function stringify(string) {
+    return '\'' + string.replace(/\\/g, '\\\\').replace(/\n/g, '\\n').replace(/'/g, '\\\'') + '\'';
+  }
+
+  function checkExpectations(testURL, passes, failures, expectedFailures, flakyFailureIndicator) {
+    expectedFailures = expectedFailures || {};
+
+    var failedDifferently = false;
+    var differentFailures = {};
+
+    var failedUnexpectedly = false;
+    var unexpectedFailures = {};
+
+    var passedUnexpectedly = false;
+    var unexpectedPasses = [];
+
+    var hasStaleFailureExpectations = false;
+    var staleFailureExpectations = [];
+
+    for (var name in failures) {
+      var message = failures[name];
+      if (name in expectedFailures) {
+        if (expectedFailures[name] != flakyFailureIndicator && message != expectedFailures[name]) {
+          failedDifferently = true;
+          differentFailures[name] = message;
+        }
+      } else {
+        failedUnexpectedly = true;
+        unexpectedFailures[name] = message;
+      }
+    }
+    for (var name in expectedFailures) {
+      if (name in passes) {
+        passedUnexpectedly = true;
+        unexpectedPasses.push(name);
+      } else if (!(name in failures)) {
+        hasStaleFailureExpectations = true;
+        staleFailureExpectations.push(name);
+      }
+    }
+
+    var message = '';
+    if (hasStaleFailureExpectations) {
+      message += 'Stale failure expectations, test no longer exists:\n';
+      message += staleFailureExpectations.map(function(name) { return '  ' + name + '\n'; }).join('');
+      message += '\n';
+    }
+    if (passedUnexpectedly) {
+      message += 'Passed unexpectedly:\n';
+      message += unexpectedPasses.map(function(name) { return '  ' + name + '\n'; }).join('');
+      message += '\n';
+    }
+    if (failedDifferently) {
+      message += 'Failed differently:\n';
+      for (var name in differentFailures) {
+        message += '  Test: ' + stringify(name) + '\n';
+        message += '  Expected: ' + stringify(expectedFailures[name]) + '\n';
+        message += '  Actual:   ' + stringify(differentFailures[name]) + '\n\n';
+      }
+      message += '\n';
+    }
+    if (failedUnexpectedly) {
+      message += 'Failed unexpectedly:\n';
+      for (var name in unexpectedFailures) {
+        message += '  Test: ' + stringify(name) + '\n';
+        message += '  Failure: ' + stringify(unexpectedFailures[name]) + '\n\n';
+      }
+      message += '\n';
+    }
+
+    return {
+      suite: [testURL],
+      description: '',
+      success: message == '',
+      log: message ? [message] : [],
+    };
+  }
+
+  function checkConfig(config) {
+    var requiredProperties = {
+      testURLList: '<Array of test URLs>',
+      skip: '<Object maping skipped test URLs with the reason for skipping>',
+      expectedFailures: '<Object mapping test URLs to expected inner test failures>',
+      flakyFailureIndicator: '<String used in expectedFailures to indicate non deterministic failure messages>',
```

Comment:
We currently don't need such a thing, adding it would make the expected failure logic even more complex for no functional gain. I'd much prefer to have just one notion of flakiness for simplicity, I suspect pass/fail is the clearer definition.